src/lustre/lustreDesugarIfBlocks.ml

```diff
@@ -0,0 +1,330 @@
+          | A.FrameBlock (pos, _, _, _)
+          | A.Body (Assert (pos, _)) 
+          | A.AnnotProperty (pos, _, _)
+          | A.AnnotMain (pos, _) -> mk_error pos (MisplacedNodeItemError ni)
+        )
+      (* We've processed everything in the if block. *)
+      | A. IfBlock (_, _, [], []) -> R.ok (trees)
+      (* shouldn't be possible *)
+      | _ -> assert false
+  ) in
+  (helper ib LhsMap.empty [])
+  
+(** Converts a tree of conditions/expressions to an ITE expression. *)
+let rec tree_to_ite pos node =
+  match node with
+    | Leaf Some expr -> expr
+    | Leaf None -> A.Ident(pos, HString.mk_hstring "ib_oracle")
+    | Node (left, cond, right) -> 
+      TernaryOp (pos, Ite, cond, tree_to_ite pos left, tree_to_ite pos right)
+
+(** Returns the type associated with a tree. *)
+let get_tree_type ctx lhs = 
+  match lhs with
+    | A.StructDef(_, [SingleIdent(_, i)]) -> (Ctx.lookup_ty ctx i) 
+    | A.StructDef(_, [ArrayDef(_, i, _)]) -> (
+      match (Ctx.lookup_ty ctx i) with
+        (* Assignment in the form of A[i] = f(i), so the RHS type is no
+           longer an array *)
+        | Some (ArrayType (_, (ty, _))) -> Some ty
+        | _ -> None
+      )
+    (* Other cases not possible *)
+    | _ -> assert false
+
+(** Fills empty spots in an ITE with oracles. *)
+let rec fill_ite_with_oracles ite ty = 
+  match ite with
+    | A.TernaryOp (pos, Ite, cond, e1, e2) -> 
+      let e1, gids1, decls1 = (
+      match e1 with
+        | TernaryOp _ as top -> fill_ite_with_oracles top ty
+        | Ident(p, s) when s = HString.mk_hstring "ib_oracle" -> 
+          let (expr, gids) = (mk_fresh_ib_oracle ty) in
+          (match expr with
+            | A.Ident (_, name) ->  expr, gids, [A.NodeVarDecl(p, (p, name, ty, ClockTrue))]
+            | _ -> assert false (* not possible*))
+        | _ -> e1, GI.empty (), []
+      ) in 
+      let e2, gids2, decls2 = (
+      match e2 with
+        | TernaryOp _ as top -> fill_ite_with_oracles top ty
+        | Ident(p, s) when s = HString.mk_hstring "ib_oracle" -> 
+          let (expr, gids) = (mk_fresh_ib_oracle ty) in
+          (match expr with
+            | A.Ident (_, name) ->  expr, gids, [A.NodeVarDecl(p, (p, name, ty, ClockTrue))]
+            | _ -> assert false (* not possible*))
+        | _ -> e2, GI.empty (), []
+      ) in
+      A.TernaryOp (pos, Ite, cond, e1, e2), GI.union gids1 gids2, decls1 @ decls2
+    (* shouldn't be possible *)
+    | _ -> assert false
+
+
+(** Helper function to determine if two trees are equal *)
+let rec trees_eq node1 node2 = match node1, node2 with
+  | Leaf (Some i), Leaf (Some j) -> (match (AH.syn_expr_equal None i j) with
+    | Ok n -> n
+    | Error _ -> false
+  )
+  | Node (l1, _, r1), Node (l2, _, r2) -> 
+    trees_eq l1 l2 && trees_eq r1 r2
+  | _ -> false
+
+  
+(** Removes redundancy from a binary tree. 
+   Currently, redundancy check doesn't quite work because
+   different positions mess with equality of leaves. *)
```

Comment:
Is that right? The description of `AH.syn_expr_equal` (used in `trees_eq`) claims positions are ignored.

Reply:
You're right. This is a deprecated comment.